dll/win32/shlwapi/appcompat.c

```diff
@@ -0,0 +1,443 @@
+               PathFindFileNameA(pszPath));
+
+    // Open the key
+    HKEY hKey;
+    const REGSAM samDesired = KEY_QUERY_VALUE | KEY_ENUMERATE_SUB_KEYS;
+    LSTATUS error = RegOpenKeyExA(HKEY_LOCAL_MACHINE, szText, 0, samDesired, &hKey);
+    if (error != ERROR_SUCCESS)
+    {
+        ERR("error: %lu\n", error);
+        return 0; // Failed
+    }
+
+    // Build the base directory
+    CHAR szBaseDir[MAX_PATH];
+    lstrcpynA(szBaseDir, pszPath, _countof(szBaseDir));
+    PathRemoveFileSpecA(szBaseDir);
+
+    // Search from the registry key
+    DWORD dwCompatFlags = 0;
+    szText[0] = ANSI_NULL; // The first attempt is for the non-subkey path
+    for (DWORD dwIndex = 0; error == ERROR_SUCCESS;)
+    {
+        // Open the sub-key
+        HKEY hSubKey;
+        error = RegOpenKeyExA(hKey, szText, 0, KEY_QUERY_VALUE, &hSubKey);
+        if (error != ERROR_SUCCESS)
+            break;
+
+        // Get the "RequiredFile" value
+        CHAR szRequired[MAX_PATH];
+        DWORD cbData = sizeof(szRequired);
+        error = SHGetValueA(hSubKey, NULL, "RequiredFile", NULL, szRequired, &cbData);
+        BOOL bValueExists = (error == ERROR_SUCCESS);
+        BOOL bRequiredFileExists = FALSE;
+        if (bValueExists) // Does the "RequiredFile" value exist?
+        {
+            // Build required file path to szText
+            PathCombineA(szText, szBaseDir, szRequired);
+            TRACE("RequiredFile: %s\n", wine_dbgstr_a(szRequired));
+            TRACE("szText: %s\n", wine_dbgstr_a(szText));
+            // Now szText is a full path
+            bRequiredFileExists = (GetFileAttributesA(szText) != INVALID_FILE_ATTRIBUTES);
+        }
+
+        // If the "RequiredFile" value doesn't exist, or the file specified by szText exists:
+        if (!bValueExists || bRequiredFileExists)
+        {
+            // Check the "Version" value if necessary
+            error = SHGetValueA(hSubKey, NULL, "Version", NULL, szText, &cbData);
+            PCSTR pszVersionPattern = ((error == ERROR_SUCCESS) ? szText : NULL);
+            TRACE("pszVersionPattern: %s\n", wine_dbgstr_a(pszVersionPattern));
+
+            // Does the pattern match?
+            if (SHLWAPI_DoesModuleVersionMatch(pszPath, pszVersionPattern))
+            {
+                // Add additional flags from the registry key
+                dwCompatFlags |= SHLWAPI_GetMappedFlags(hSubKey, g_appCompatFlagMaps,
+                                                        _countof(g_appCompatFlagMaps));
+            }
+        }
+
+        // Close the sub-key
+        RegCloseKey(hSubKey);
+
+        // Proceed to the next sub-key
+        ++dwIndex;
+        error = RegEnumKeyA(hKey, dwIndex, szText, _countof(szText));
+    }
+
+    // Close the key
+    RegCloseKey(hKey);
+
+    return dwCompatFlags;
+}
+
+static VOID
+SHLWAPI_InitAppCompat(VOID)
+{
+    if (GetProcessVersion(0) >= 0x50000)
```

Comment:
```suggestion
    if (GetProcessVersion(0) >= MAKELONG(0, 5))
```
(minor == 0, major == 5)